<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Annotate genes with GO terms and measure their activities — map_go_annot • ParetoTI</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Annotate genes with GO terms and measure their activities — map_go_annot" />

<meta property="og:description" content="map_go_annot() annotates genes with Gene Ontology terms including less specific terms. It automatically loads correct OrgDb package given taxonomy_id. Then it selects GOALL annotations for genes provided in keys, filter by ontology branch and evidence codes. Finally it maps GO term names using GO.db and returns annotations in list and data.table formats. This function can be used to retrieve other gene annotations from OrgDb (e.g OMIM, PFAM) but those will not be mapped to readable names.
filter_go_annot() filters GO annotations by ontology file, GO branch, number of genes annotated
measure_activity() annotates genes with GO terms and measures their activities
load_go_ontology() load GO slim ontology file and return file path
map_gwas_annot() Load known GWAS loci for all phenotypes in GWAS Catalog. For each the keys find which phenotypes it is associated with using genes mapped by GWAS Catalog.
map_promoter_tf() load GO slim ontology file and return file path" />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ParetoTI</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.7</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/Hepatocyte_example.html">Example of using archetypal analysis to find representative cells &amp; describe heterogeniety in hepatocyte population between those archetypes</a>
    </li>
  </ul>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Annotate genes with GO terms and measure their activities</h1>
    
    <div class="hidden name"><code>measure_activity.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>map_go_annot() annotates genes with Gene Ontology terms including less specific terms. It automatically loads correct OrgDb package given <code>taxonomy_id</code>. Then it selects GOALL annotations for genes provided in <code>keys</code>, filter by ontology branch and evidence codes. Finally it maps GO term names using GO.db and returns annotations in list and data.table formats. This function can be used to retrieve other gene annotations from OrgDb (e.g OMIM, PFAM) but those will not be mapped to readable names.</p>
<p>filter_go_annot() filters GO annotations by ontology file, GO branch, number of genes annotated</p>
<p>measure_activity() annotates genes with GO terms and measures their activities</p>
<p>load_go_ontology() load GO slim ontology file and return file path</p>
<p>map_gwas_annot() Load known GWAS loci for all phenotypes in GWAS Catalog. For each the <code>keys</code> find which phenotypes it is associated with using genes mapped by GWAS Catalog.</p>
<p>map_promoter_tf() load GO slim ontology file and return file path</p>
    
    </div>

    <pre class="usage"><span class='fu'>map_go_annot</span>(<span class='kw'>taxonomy_id</span> <span class='kw'>=</span> <span class='fl'>9606</span>, <span class='kw'>keys</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='st'>"TP53"</span>, <span class='st'>"ALB"</span>),
  <span class='kw'>columns</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='st'>"GOALL"</span>), <span class='kw'>keytype</span> <span class='kw'>=</span> <span class='st'>"ALIAS"</span>, <span class='kw'>ontology_type</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='st'>"BP"</span>,
  <span class='st'>"MF"</span>, <span class='st'>"CC"</span>), <span class='kw'>evidence_code</span> <span class='kw'>=</span> <span class='st'>"all"</span>, <span class='kw'>localHub</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>return_org_db</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>ann_hub_cache</span> <span class='kw'>=</span> <span class='kw pkg'>AnnotationHub</span><span class='kw ns'>::</span><span class='fu'><a href='https://www.rdocumentation.org/packages/AnnotationHub/topics/getAnnotationHubOption'>getAnnotationHubOption</a></span>(<span class='st'>"CACHE"</span>))

<span class='fu'>filter_go_annot</span>(<span class='no'>annot</span>, <span class='kw'>ontology_file</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>lower</span> <span class='kw'>=</span> <span class='fl'>1</span>, <span class='kw'>upper</span> <span class='kw'>=</span> <span class='fl'>Inf</span>,
  <span class='kw'>ontology_type</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='st'>"BP"</span>, <span class='st'>"MF"</span>, <span class='st'>"CC"</span>))

<span class='fu'>measure_activity</span>(<span class='no'>expr_mat</span>, <span class='kw'>which</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='st'>"BP"</span>, <span class='st'>"MF"</span>, <span class='st'>"CC"</span>, <span class='st'>"gwas"</span>,
  <span class='st'>"promoter_TF"</span>, <span class='st'>"custom"</span>)[<span class='fl'>1</span>], <span class='kw'>activity_method</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='st'>"AUCell"</span>,
  <span class='st'>"pseudoinverse"</span>)[<span class='fl'>1</span>], <span class='kw'>keys</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/colnames'>rownames</a></span>(<span class='no'>expr_mat</span>), <span class='kw'>keytype</span> <span class='kw'>=</span> <span class='st'>"ALIAS"</span>,
  <span class='kw'>ontology_file</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>taxonomy_id</span> <span class='kw'>=</span> <span class='fl'>10090</span>, <span class='kw'>evidence_code</span> <span class='kw'>=</span> <span class='st'>"all"</span>,
  <span class='kw'>localHub</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>ann_hub_cache</span> <span class='kw'>=</span> <span class='kw pkg'>AnnotationHub</span><span class='kw ns'>::</span><span class='fu'><a href='https://www.rdocumentation.org/packages/AnnotationHub/topics/getAnnotationHubOption'>getAnnotationHubOption</a></span>(<span class='st'>"CACHE"</span>),
  <span class='kw'>lower</span> <span class='kw'>=</span> <span class='fl'>1</span>, <span class='kw'>upper</span> <span class='kw'>=</span> <span class='fl'>Inf</span>, <span class='kw'>return_as_matrix</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>annot_dt</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>assay_name</span> <span class='kw'>=</span> <span class='st'>"logcounts"</span>,
  <span class='kw'>aucell_options</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/list'>list</a></span>(<span class='kw'>aucMaxRank</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/nrow'>nrow</a></span>(<span class='no'>expr_mat</span>) * <span class='fl'>0.05</span>, <span class='kw'>binary</span> <span class='kw'>=</span> <span class='no'>F</span>,
  <span class='kw'>nCores</span> <span class='kw'>=</span> <span class='fl'>3</span>, <span class='kw'>plotStats</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>))

<span class='fu'>load_go_ontology</span>(<span class='kw'>ont_dir</span> <span class='kw'>=</span> <span class='st'>"./data/"</span>, <span class='kw'>ont_file</span> <span class='kw'>=</span> <span class='st'>"goslim_generic.obo"</span>)

<span class='fu'>map_gwas_annot</span>(<span class='kw'>taxonomy_id</span> <span class='kw'>=</span> <span class='fl'>9606</span>, <span class='kw'>keys</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='st'>"TP53"</span>, <span class='st'>"ZZZ3"</span>),
  <span class='kw'>keytype</span> <span class='kw'>=</span> <span class='st'>"SYMBOL"</span>, <span class='kw'>localHub</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>ann_hub_cache</span> <span class='kw'>=</span> <span class='kw pkg'>AnnotationHub</span><span class='kw ns'>::</span><span class='fu'><a href='https://www.rdocumentation.org/packages/AnnotationHub/topics/getAnnotationHubOption'>getAnnotationHubOption</a></span>(<span class='st'>"CACHE"</span>),
  <span class='kw'>gwas_url</span> <span class='kw'>=</span> <span class='st'>"https://www.ebi.ac.uk/gwas/api/search/downloads/alternative"</span>,
  <span class='kw'>gwas_file</span> <span class='kw'>=</span> <span class='st'>"gwas_catalog_v1.0.2-associations_e93_r2019-01-31.tsv"</span>,
  <span class='kw'>lower</span> <span class='kw'>=</span> <span class='fl'>1</span>, <span class='kw'>upper</span> <span class='kw'>=</span> <span class='fl'>Inf</span>)

<span class='fu'>map_promoter_tf</span>(<span class='kw'>keys</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='st'>"TP53"</span>, <span class='st'>"ZZZ3"</span>), <span class='kw'>keytype</span> <span class='kw'>=</span> <span class='st'>"SYMBOL"</span>,
  <span class='kw'>lower</span> <span class='kw'>=</span> <span class='fl'>1</span>, <span class='kw'>upper</span> <span class='kw'>=</span> <span class='fl'>Inf</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>taxonomy_id</th>
      <td><p>Taxonomy id of your species. 9606 is human, 10090 is mouse. Find your species id using taxonomy search by UniProt: https://www.uniprot.org/taxonomy/.</p></td>
    </tr>
    <tr>
      <th>keys</th>
      <td><p>gene or protein identifiers of <code>keytype</code></p></td>
    </tr>
    <tr>
      <th>columns</th>
      <td><p>columns to retrieve. Default is GOALL: GO Identifiers (includes less specific terms), colsAndKeytypes.</p></td>
    </tr>
    <tr>
      <th>keytype</th>
      <td><p>type of <code>keys</code> identifiers. Use <code>keytypes(map_go_annot(taxonomy_id = 9606, return_org_db = T))</code> to find which keytypes are available.</p></td>
    </tr>
    <tr>
      <th>ontology_type</th>
      <td><p>specify which branch of gene ontology to retrieve annotations from.</p></td>
    </tr>
    <tr>
      <th>evidence_code</th>
      <td><p>specify which evidence should have been used to annotate genes with GO terms. Use <code>keys(map_go_annot(..., return_org_db = T), "EVIDENCE")</code> to find which codes exist. See http://www.geneontology.org/page/guide-go-evidence-codes for explanations and details.</p></td>
    </tr>
    <tr>
      <th>localHub</th>
      <td><p>set localHub = FALSE for working offline. Details <a href='https://www.rdocumentation.org/packages/AnnotationHub/topics/AnnotationHub-class'>AnnotationHub</a>.</p></td>
    </tr>
    <tr>
      <th>return_org_db</th>
      <td><p>return OrgDb database link instead of gene annotations.</p></td>
    </tr>
    <tr>
      <th>ann_hub_cache</th>
      <td><p>If needed, set location of local AnnotationHub cache to specific directory by providing it here. To view where cache is currently located use getAnnotationHubOption("CACHE")</p></td>
    </tr>
    <tr>
      <th>annot</th>
      <td><p>GO annotations, output of map_go_annot()</p></td>
    </tr>
    <tr>
      <th>ontology_file</th>
      <td><p>path (url) to ontology file in obo format</p></td>
    </tr>
    <tr>
      <th>lower</th>
      <td><p>lower limit on size of gene sets (inclusive)</p></td>
    </tr>
    <tr>
      <th>upper</th>
      <td><p>upper limit on size of gene sets (exclusive)</p></td>
    </tr>
    <tr>
      <th>expr_mat</th>
      <td><p>expression matrix (genes in rows, cells in columns) or one of: dgCMatrix, ExpressionSet, and SummarizedExperiment or SingleCellExperiment both of which require assay_name.</p></td>
    </tr>
    <tr>
      <th>which</th>
      <td><p>which set activities to measure? Currently implemented "BP", "MF", "CC" gene ontology subsets. Use "gwas" for constructing gene sets with gwas_catalog v1.0.2. GWAS option works only with human hgnc identifiers. Use "promoter_TF" to look at TF activities measured by mapping TF-&gt;target gene regulation via co-occurence of multiple motifs for the same TF in promoters (courtesy of Krzysztof Polanski who generated this resource). TF data is added supplied with the package and works only for human data, only SYMBOL and ENSEMBL can be used as keytype. Use "custom" to provide your own gene sets in <code>annot_dt</code> table.</p></td>
    </tr>
    <tr>
      <th>activity_method</th>
      <td><p>find activities using <a href='find_set_activity_AUCell.html'>find_set_activity_AUCell</a> or <a href='find_set_activity_AUCell.html'>find_set_activity_pseudoinv</a>?</p></td>
    </tr>
    <tr>
      <th>return_as_matrix</th>
      <td><p>return matrix (TRUE) or data.table (FALSE)</p></td>
    </tr>
    <tr>
      <th>annot_dt</th>
      <td><p>data.table, data.frame or matrix containing custom gene set annotations. The 1st column should contain gene set identifier or name, the 2nd column should contain gene identifiers matching <code>keys</code>.</p></td>
    </tr>
    <tr>
      <th>assay_name</th>
      <td><p>name of assay in SummarizedExperiment or SingleCellExperiment, normally counts or logcounts</p></td>
    </tr>
    <tr>
      <th>aucell_options</th>
      <td><p>additional options specific to AUCell, details: <a href='find_set_activity_AUCell.html'>find_set_activity_AUCell</a></p></td>
    </tr>
    <tr>
      <th>ont_dir</th>
      <td><p>directory where to save ontology</p></td>
    </tr>
    <tr>
      <th>ont_file</th>
      <td><p>which ontology file to load?</p></td>
    </tr>
    <tr>
      <th>gwas_url</th>
      <td><p>where to download annotations? Useful for other versions of GWAS Catalog.</p></td>
    </tr>
    <tr>
      <th>gwas_file</th>
      <td><p>name of the file where to store GWAS Catalog data locally (in ann_hub_cache directory).</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>list (S3 class ParetoTI_annot) containing GO annotations in list and data.table formats, and OrgDb database link.</p>
    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>GO consortium annotates genes with the most specific terms so there is a need to propagate annotations to less specific parent terms. For example, "apoptotic process in response to mitochondrial fragmentation" (GO:0140208) is an "apoptotic process" (GO:0006915). Conceptually, if gene functions in the first it also functions in the second but it is only directly annotated by the first term. So, using GO annotations for functional enrichment requires propagating annotations to less specific terms.</p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='no'>annot</span> <span class='kw'>=</span> <span class='fu'>map_go_annot</span>(<span class='kw'>taxonomy_id</span> <span class='kw'>=</span> <span class='fl'>9606</span>,
                     <span class='kw'>keys</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='st'>"TP53"</span>, <span class='st'>"ALB"</span>, <span class='st'>"GPX1"</span>),
                     <span class='kw'>columns</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='st'>"GOALL"</span>), <span class='kw'>keytype</span> <span class='kw'>=</span> <span class='st'>"ALIAS"</span>,
                     <span class='kw'>ontology_type</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='st'>"BP"</span>, <span class='st'>"MF"</span>, <span class='st'>"CC"</span>))</div><div class='output co'>#&gt; <span class='message'>snapshotDate(): 2018-10-24</span></div><div class='output co'>#&gt; <span class='message'>downloading 0 resources</span></div><div class='output co'>#&gt; <span class='message'>loading from cache </span>
#&gt; <span class='message'>    ‘/Users/vk7//.AnnotationHub/72902’</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: AnnotationDbi</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: stats4</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: BiocGenerics</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: parallel</span></div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>Attaching package: ‘BiocGenerics’</span></div><div class='output co'>#&gt; <span class='message'>The following objects are masked from ‘package:parallel’:</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,</span>
#&gt; <span class='message'>    clusterExport, clusterMap, parApply, parCapply, parLapply,</span>
#&gt; <span class='message'>    parLapplyLB, parRapply, parSapply, parSapplyLB</span></div><div class='output co'>#&gt; <span class='message'>The following objects are masked from ‘package:stats’:</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    IQR, mad, sd, var, xtabs</span></div><div class='output co'>#&gt; <span class='message'>The following objects are masked from ‘package:base’:</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    anyDuplicated, append, as.data.frame, basename, cbind, colMeans,</span>
#&gt; <span class='message'>    colnames, colSums, dirname, do.call, duplicated, eval, evalq,</span>
#&gt; <span class='message'>    Filter, Find, get, grep, grepl, intersect, is.unsorted, lapply,</span>
#&gt; <span class='message'>    lengths, Map, mapply, match, mget, order, paste, pmax, pmax.int,</span>
#&gt; <span class='message'>    pmin, pmin.int, Position, rank, rbind, Reduce, rowMeans, rownames,</span>
#&gt; <span class='message'>    rowSums, sapply, setdiff, sort, table, tapply, union, unique,</span>
#&gt; <span class='message'>    unsplit, which, which.max, which.min</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: Biobase</span></div><div class='output co'>#&gt; <span class='message'>Welcome to Bioconductor</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    Vignettes contain introductory material; view with</span>
#&gt; <span class='message'>    'browseVignettes()'. To cite Bioconductor, see</span>
#&gt; <span class='message'>    'citation("Biobase")', and for packages 'citation("pkgname")'.</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: IRanges</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: S4Vectors</span></div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>Attaching package: ‘S4Vectors’</span></div><div class='output co'>#&gt; <span class='message'>The following objects are masked from ‘package:data.table’:</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    first, second</span></div><div class='output co'>#&gt; <span class='message'>The following object is masked from ‘package:base’:</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    expand.grid</span></div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>Attaching package: ‘IRanges’</span></div><div class='output co'>#&gt; <span class='message'>The following object is masked from ‘package:data.table’:</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    shift</span></div><div class='input'><span class='no'>annot2</span> <span class='kw'>=</span> <span class='fu'>filter_go_annot</span>(<span class='no'>annot</span>, <span class='kw'>ontology_file</span> <span class='kw'>=</span> <span class='st'>"http://www.geneontology.org/ontology/subsets/goslim_generic.obo"</span>,
           <span class='kw'>lower</span> <span class='kw'>=</span> <span class='fl'>50</span>, <span class='kw'>upper</span> <span class='kw'>=</span> <span class='fl'>2000</span>, <span class='kw'>ontology_type</span> <span class='kw'>=</span> <span class='st'>"BP"</span>)
<span class='no'>activ</span> <span class='kw'>=</span> <span class='fu'>measure_activity</span>(<span class='no'>expr_mat</span>, <span class='kw'>which</span> <span class='kw'>=</span> <span class='st'>"BP"</span>,
                         <span class='kw'>taxonomy_id</span> <span class='kw'>=</span> <span class='fl'>9606</span>, <span class='kw'>keytype</span> <span class='kw'>=</span> <span class='st'>"ALIAS"</span>,
                         <span class='kw'>ontology_file</span> <span class='kw'>=</span> <span class='fu'>load_go_ontology</span>(<span class='st'>"./data/"</span>,
                                                  <span class='st'>"goslim_generic.obo"</span>))</div><div class='output co'>#&gt; <span class='message'>snapshotDate(): 2018-10-24</span></div><div class='output co'>#&gt; <span class='message'>downloading 0 resources</span></div><div class='output co'>#&gt; <span class='message'>loading from cache </span>
#&gt; <span class='message'>    ‘/Users/vk7//.AnnotationHub/72902’</span></div><div class='output co'>#&gt; <span class='error'>Error in rownames(expr_mat): object 'expr_mat' not found</span></div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Vitalii Kleshchevnikov.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
   </div>

  

  </body>
</html>

